File: examples/src/queries/multi_resolution.cc
#include <wavemap/data_structure/volumetric/hashed_wavelet_octree.h>
#include <wavemap/indexing/index_conversions.h>

#include "wavemap_examples/common.h"

using namespace wavemap;
int main(int, char**) {
  // Declare a map pointer for illustration purposes
  // NOTE: See the other tutorials on how to load maps from files or ROS topics,
  //       such as the map topic published by the wavemap ROS server.
  HashedWaveletOctree::Ptr map;

  // Define the center point and the minimum width of the octree cell to query
  const Point3D query_point = Point3D::Zero();
  const FloatingPoint query_min_cell_width = 0.5f;  // in meters

  // Convert it to an octree node index
  const FloatingPoint map_min_cell_width = map->getMinCellWidth();
  const IndexElement query_height = convert::cellWidthToHeight(
      query_min_cell_width, 1.f / map_min_cell_width);
  const OctreeIndex query_index =
      convert::pointToNodeIndex(query_point, map_min_cell_width, query_height);

  // Query the map
  const FloatingPoint occupancy_log_odds = map->getCellValue(query_index);
  examples::doSomething(occupancy_log_odds);
}

File: examples/python/queries/multi_resolution.py
import numpy as np
import pywavemap as wave
import _dummy_objects

# Load a map
your_map = _dummy_objects.example_map()

# Define the center point and the minimum width of your region of interest
query_point = np.array([0.4, 0.5, 0.6])
query_min_cell_width = 0.5  # in meters

# Compute the index of the smallest node that covers it completely
query_height = wave.convert.cell_width_to_height(query_min_cell_width,
                                                 your_map.min_cell_width)
query_index = wave.convert.point_to_node_index(query_point,
                                               your_map.min_cell_width,
                                               query_height)

# Query the node's average occupancy
occupancy_log_odds = your_map.getCellValue(query_index)
print(occupancy_log_odds)
